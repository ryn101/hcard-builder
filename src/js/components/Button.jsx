import React from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames';

const validElementTags = ['span', 'div', 'button'];

function Button({
    children, colour, as, ...props
}) {
    const ButtonTag = as;

    return (
        // eslint-disable-next-line react/jsx-props-no-spreading
        <ButtonTag className={classNames('button', `button--${colour}`)} {...props}>
            { children }
        </ButtonTag>
    );
}

// noinspection JSUnusedGlobalSymbols
Button.propTypes = {
    children: PropTypes.oneOfType([
        PropTypes.element,
        PropTypes.string,
    ]).isRequired,
    colour: PropTypes.string,
    as: PropTypes.oneOf(validElementTags),
};

Button.defaultProps = {
    colour: 'grey',
    as: 'button',
};

export default Button;
