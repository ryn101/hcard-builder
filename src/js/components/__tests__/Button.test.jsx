import React from 'react';
import ReactDOM from 'react-dom';
import { shallow, configure } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import Button from '../Button';

configure({ adapter: new Adapter() });

it('renders the button component without crashing', () => {
    const div = document.createElement('div');
    ReactDOM.render(
        <Button>Button Text</Button>,
        div,
    );
    ReactDOM.unmountComponentAtNode(div);
});

describe('once rendered', () => {
    const buttonText = 'Button Text';
    const buttonColour = 'blue';
    const getButtonColour = (colour) => `button--${colour}`;
    let buttonNode;

    beforeEach(() => {
        buttonNode = shallow(
            <Button colour={buttonColour}>{buttonText}</Button>,
        );
    });

    it('renders with state correctly', () => {
        expect(buttonNode).toMatchSnapshot();
    });

    it('defaults to the grey colour', () => {
        const greyColour = 'grey';
        buttonNode.setProps({ colour: greyColour });

        expect(buttonNode.hasClass(getButtonColour(greyColour))).toBeTruthy();
    });

    it('contains the correct props', () => {
        expect(buttonNode.text()).toEqual(buttonText);
        expect(buttonNode.hasClass(getButtonColour(buttonColour))).toBeTruthy();
    });

    it('supports different and rejects unsupported tags', () => {
        const tag = 'span';

        buttonNode.setProps({ as: tag });
        expect(buttonNode.name()).toEqual(tag);

        expect(() => {
            buttonNode.setProps({ as: 'unsupportedtag' });
        }).toThrow();
    });
});
