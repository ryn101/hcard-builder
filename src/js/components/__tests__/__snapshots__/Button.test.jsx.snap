// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`once rendered renders with state correctly 1`] = `
<button
  className="button button--blue"
>
  Button Text
</button>
`;
